### PR TITLE
Filter unknown input types

### PR DESCRIPTION
**Pull request by sjensen85**
This is attempting to address #16 with @adamkl tests and a couple added tests. I feel like the approach taken is too verbose and probably done incredibly wrong :smile: but wanted to get this out there for discussion.

### CODE REVIEW

File: src/rules/NoUnknownInputTypes.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  return {
+    OperationDefinition(node) {
+      if (node.variableDefinitions.length > 0) {
+        node.variableDefinitions.forEach(def => {
+          if (def.type.type) {
+            const type = def.type.type;
+            if (!context._schema._typeMap[type.name.value]) {
+              context.reportError(
+                new _error.GraphQLError(
+                  noUnknownInputTypeMessage(node.name.value, type),
```
Comment (Viper512) [inline src/rules/NoUnknownInputTypes.js:23]: should this be using type.name.value?

Comment (sjensen85) [reply src/rules/NoUnknownInputTypes.js:23]: Maybe not. What do you suggest?

Comment (adamkl) [reply src/rules/NoUnknownInputTypes.js:23]: @Viper512 (Corey G.) pulled in the latest `graphql-mask` and ran into another issue. I'm not sure if its related to this change, or something else (we still have an issue with the mutations, but a different error this time). 

I think he's been fiddling around with the code in `graphql-mask` to see if he could get it to work, but I don't know that we've verified anything against the current unit tests.  We probably need to dig a little deeper.

Comment (sjensen85) [reply src/rules/NoUnknownInputTypes.js:23]: Yeah, I'm not sure the best way to work with the objects here. I was surprised it was merged without comment. 😄 

The `type.name.value` is the declared type name and key for `typeMap`. This was the only way I could figure out how to access them at the time. There most likely is a better way. 

What error are you receiving now?

Comment (Viper512) [reply src/rules/NoUnknownInputTypes.js:23]: with node.name.value, i'm receiving the following error : 
(node:11740) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): TypeError: Cannot read property 'value' of undefined
(node:11740) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

With type.name.value, it's working fine for me.

Which makes me think that Line 23 has a typo on it.

Comment (sjensen85) [reply src/rules/NoUnknownInputTypes.js:23]: Hmm.. I feel like every operation should have a name, but could be wrong. Same goes for types.

What do the query and schema look like? Can you provide a reproduction?

@brysgo may need to shed some light here.

**Comment by adamkl** [comment]
I’ll try to work out a repro with @Viper512 mid next week or so.

**Comment by sjensen85** [comment]
Thanks @adamkl @Viper512

**Comment by brysgo** [comment]
Other than saying we likely can't assume that the node will have a name since the name may have been removed by a validation error I don't have any light to shed.

**Comment by brysgo** [comment]
I would also suggest if you are using this library that you run graphql validation on a combined schema before you mask a query. The error you have could be because of a legitimate validation error.

**Comment by brysgo** [comment]
^^ @adamkl @viper512

**Comment by sjensen85** [comment]
Not knowing enough about graphql validation, I didn't think about that. 

I can not use the `node.name.value`, I included it explicitly for error message reasons.

@adamkl @Viper512 @brysgo I can make a follow up PR without.

**Comment by sjensen85** [comment]
@brysgo we may need to update the `UnsupportedOperation` rule - it uses the same key. Should I update that in my PR?

**Comment by brysgo** [comment]
Don't do anything yet, let's wait on a failing test from @viper512

**Comment by brysgo** [comment]
There is a possibility that manual graphql validation before graphql mask will solve this problem.

**Pull request merged**